common/Vectorize.py

```diff
@@ -0,0 +1,307 @@
+
+        Also adds a dummy node which is connected to all inputs
+        if the graph is not connected.
+
+        Arguments:
+            graph (Graph object) : Graph object to be converted to networkx
+            include_edge_attrs (str) : case insensitive string to denote 
+                whether to include edge attributes of input edge from 
+                _EDGE_ATTRS in feature, if "true" then they are included.
+            include_node_attrs (str) : case insensitive string to denote 
+                whether to include node attributes from _NODE_ATTRS in feature,
+                if "true" then they are included.
+
+        Returns:
+            networkx graph corresponding to the converted Graph object        
+        """
+
+        # Initializing networkx graph
+        networkx_graph = networkx.Graph(
+            model_name = graph.model_name, category = graph.category,
+            sub_category = graph.sub_category, source = graph.source
+            )
+
+        input_edges = dict()
+
+        # Adding edges to the graph
+        for src_node_index in graph.adj_list.keys():
+            for [edge_index, dest_node_index] in graph.adj_list[src_node_index]:
+                networkx_graph.add_edge(
+                    src_node_index, dest_node_index,
+                    tensor_shape = graph.edges[edge_index].tensor_shape,
+                    tensor_type = graph.edges[edge_index].tensor_type)
+                
+                if dest_node_index not in input_edges:
+                    input_edges.update({dest_node_index : []})
+            
+                input_edges[dest_node_index].append(graph.edges[edge_index])
+        
+        # Adding all nodes to the graph, and building features
+        for index in range(len(graph.nodes)):
+            features = list()
+
+            if include_node_attrs.lower() == "true":
+                for node_attr in self._NODE_ATTRS:
+                    features.append(str(getattr(graph.nodes[index], node_attr)))
+
+            if include_edge_attrs.lower() == "true" and index in input_edges:
+                for input_edge in input_edges[index]:
+                    for edge_attr in self._EDGE_ATTRS:
+                        features.append(str(getattr(input_edge, edge_attr)))
+
+            concat_feature = " ".join(features)
+
+            networkx_graph.add_node(
+                index, feature = concat_feature)
+
+        # If graph is not connected, introduce a dummy node and connect all 
+        # inputs to it to make it connected.
+        if not networkx.algorithms.is_connected(networkx_graph):
+            max_node_num = max(networkx_graph.nodes())
+
+            networkx_graph.add_node(max_node_num + 1, feature = "Dummy")
+
+            for index in graph.start_node_indices:
+                networkx_graph.add_edge(max_node_num + 1, index)
+
+        return networkx_graph
+
+    def get_graph2vec_embeddings(self, include_edge_attrs = False, 
```

Comment:
Are the embeddings also stored in the db? I guess then when we add a new model, we can trigger a re-training automatically (using Spanner queues and receivers) to get new embeddings since the training is on the entire set of existing models - not immediately required, something to think about for future.

Reply:
Added 'embeddings' field in Models table and added function to load the embeddings.